clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -520,9 +535,15 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
         }
       }
     } else if (const auto *FPTy = dyn_cast<FunctionProtoType>(Ty)) {
-      if (FPTy->isVariadic() && SemaRef.getLangOpts().SYCLIsDevice)
-	SemaRef.SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_sycl_restrict)
-            << Sema::KernelCallVariadicFunction;
+      if (FPTy->isVariadic() && SemaRef.getLangOpts().SYCLIsDevice) {
+        // Check if it is a printf declaration in SPIR-V friendly format.
+        const auto *VD = Decl ? Decl->getDecl() : nullptr;
+        const auto *FDecl = VD ? dyn_cast<FunctionDecl>(VD) : nullptr;
+        // FIXME: is there better solution than checking for function name?
+        if (!FDecl || !FDecl->getName().equals("__spirv_ocl_printf"))
```

Comment:
Is it possible to make this condition a little bit more obvious?
For example: diagnostic is needed when we have a `FunctionDecl` and it has name which is not equal to "__spirv_ocl_printf", i.e. :
`if (FDecl && !FDecl->getName().equals("__spirv_ocl_printf"))`

Reply:
`(!A || !B) != (A && !B)`.

> diagnostic is needed when we have a FunctionDecl and it has name which is not equal to "__spirv_ocl_printf

This is not entirely correct: diagnostic is also needed when we don't have a FunctionDecl.

I could try something like:
```c++
bool IsSpirvFriendlyPrintf = (FDecl && FDecl->getName().equals("__spirv_ocl_printf");
if (!IsSpirvFriendlyPrintf)
  // emit diag
```

What do you think?

Reply:
Could you please explain why diagnostic is also needed when we don't have a FunctionDecl?